File: solutions/0047_Permutations_II/dfs.kt
class Solution {

    var checked = booleanArrayOf()
    var ret = mutableListOf<List<Int>>()

    fun permuteUnique(nums: IntArray): List<List<Int>> {
        var arr = intArrayOf()
        checked = BooleanArray(nums.size)
        nums.sort()
        permUnit(nums, arr)
        return ret
    }

    fun permUnit(nums: IntArray, arr: IntArray) {
        if (arr.size >= nums.size) {
            ret.add(arr.toList())
            return
        }
        for (i in 0..nums.size - 1) {
            if (checked[i]) continue;
            // remove duplicates
            if (i > 0 && nums[i-1] == nums[i] && !checked[i-1]) continue
            checked[i] = true
            permUnit(nums, arr + intArrayOf(nums[i]))
            checked[i] = false
        }
    }
}

File: solutions/0306_Additive_Number/dfs.py
class Solution:
    def isAdditiveNumber(self, num: str) -> bool:
        if len(num) <= 2:
            return False
        self.ret = False
        self.dfs(num, [], 0)
        return self.ret

    def dfs(self, num, arr, k):
        if self.ret or k == len(num):
            if len(arr) >= 3 and arr[-3] + arr[-2] == arr[-1]:
                self.ret = True
            return
        if num[k] == '0':
            self.dfs(num, arr + [0], k + 1)
            return
        for p in range(k + 1, len(num) + 1):
            if len(arr) < 2 or arr[-2] + arr[-1] == int(num[k:p]):
                self.dfs(num, arr + [int(num[k:p])], p)
